File: frontend/src/model/validate.test.ts
import { describe, expect, it } from "vitest";
import { validateProjectModel } from "./validate";
import type { ModelFeature, ModelReceiver } from "./types";
import { SELF_INTERSECTION_POINT_LIMIT } from "./self-intersection";

const validSource: ModelFeature = {
  id: "src-1",
  kind: "source",
  sourceType: "point",
  geometry: { type: "Point", coordinates: [10, 51] },
};

const validBuilding: ModelFeature = {
  id: "bld-1",
  kind: "building",
  heightM: 12,
  geometry: {
    type: "Polygon",
    coordinates: [
      [
        [0, 0],
        [1, 0],
        [1, 1],
        [0, 1],
        [0, 0],
      ],
    ],
  },
};

const validBarrier: ModelFeature = {
  id: "bar-1",
  kind: "barrier",
  heightM: 3,
  geometry: {
    type: "LineString",
    coordinates: [
      [0, 0],
      [1, 1],
    ],
  },
};

const validReceiver: ModelReceiver = {
  id: "rcv-1",
  heightM: 4,
  geometry: { type: "Point", coordinates: [2, 3] },
};

describe("validateProjectModel", () => {
  it("valid model returns valid=true with no errors", () => {
    const report = validateProjectModel(
      [validSource, validBuilding, validBarrier],
      [],
    );
    expect(report.valid).toBe(true);
    expect(report.errors).toHaveLength(0);
  });

  it("empty model produces an error", () => {
    const report = validateProjectModel([], []);
    expect(report.valid).toBe(false);
    expect(report.errors[0]?.code).toBe("model.empty");
  });

  it("source without source_type produces error", () => {
    const bad: ModelFeature = {
      id: "s1",
      kind: "source",
      geometry: { type: "Point", coordinates: [0, 0] },
    };
    const report = validateProjectModel([bad], []);
    expect(report.errors.some((e) => e.code === "source.type.required")).toBe(
      true,
    );
  });

  it("source with wrong geometry produces error", () => {
    const bad: ModelFeature = {
      id: "s1",
      kind: "source",
      sourceType: "point",
      geometry: {
        type: "Polygon",
        coordinates: [
          [
            [0, 0],
            [1, 0],
            [1, 1],
            [0, 1],
            [0, 0],
          ],
        ],
      },
    };
    const report = validateProjectModel([bad], []);
    expect(
      report.errors.some((e) => e.code === "source.geometry.mismatch"),
    ).toBe(true);
  });

  it("building without height produces error", () => {
    const bad: ModelFeature = {
      id: "b1",
      kind: "building",
      geometry: {
        type: "Polygon",
        coordinates: [
          [
            [0, 0],
            [1, 0],
            [1, 1],
            [0, 1],
            [0, 0],
          ],
        ],
      },
    };
    const report = validateProjectModel([bad], []);
    expect(
      report.errors.some((e) => e.code === "building.height.required"),
    ).toBe(true);
  });

  it("building with negative height produces error", () => {
    const bad: ModelFeature = {
      id: "b1",
      kind: "building",
      heightM: -5,
      geometry: {
        type: "Polygon",
        coordinates: [
          [
            [0, 0],
            [1, 0],
            [1, 1],
            [0, 1],
            [0, 0],
          ],
        ],
      },
    };
    const report = validateProjectModel([bad], []);
    expect(
      report.errors.some((e) => e.code === "building.height.invalid"),
    ).toBe(true);
  });

  it("barrier without height produces error", () => {
    const bad: ModelFeature = {
      id: "br1",
      kind: "barrier",
      geometry: {
        type: "LineString",
        coordinates: [
          [0, 0],
          [1, 1],
        ],
      },
    };
    const report = validateProjectModel([bad], []);
    expect(
      report.errors.some((e) => e.code === "barrier.height.required"),
    ).toBe(true);
  });

  it("building with non-polygon geometry produces error", () => {
    const bad: ModelFeature = {
      id: "b1",
      kind: "building",
      heightM: 10,
      geometry: { type: "Point", coordinates: [0, 0] },
    };
    const report = validateProjectModel([bad], []);
    expect(
      report.errors.some((e) => e.code === "building.geometry.invalid"),
    ).toBe(true);
  });

  it("duplicate IDs produce error", () => {
    const a = { ...validSource };
    const b = { ...validBuilding, id: "src-1" };
    const report = validateProjectModel([a, b], []);
    expect(
      report.errors.some(
        (e) =>
          e.code === "feature.id.duplicate" ||
          e.code === "receiver.id.duplicate",
      ),
    ).toBe(true);
  });

  it("receiver validation accepts finite coordinates and positive height", () => {
    const report = validateProjectModel([validSource], [validReceiver]);
    expect(report.valid).toBe(true);
  });

  it("receiver validation rejects invalid height", () => {
    const report = validateProjectModel(
      [validSource],
      [{ ...validReceiver, heightM: 0 }],
    );
    expect(
      report.errors.some((e) => e.code === "receiver.height.invalid"),
    ).toBe(true);
  });

  it("receiver validation rejects invalid coordinates", () => {
    const report = validateProjectModel(
      [validSource],
      [
        {
          ...validReceiver,
          geometry: { type: "Point", coordinates: [Number.NaN, 3] },
        },
      ],
    );
    expect(
      report.errors.some((e) => e.code === "receiver.coordinates.invalid"),
    ).toBe(true);
  });

  it("receiver ids must not collide with feature ids", () => {
    const report = validateProjectModel(
      [validSource],
      [{ ...validReceiver, id: validSource.id }],
    );
    expect(report.errors.some((e) => e.featureId === validSource.id)).toBe(
      true,
    );
  });

  it("warns when imported source acoustics require review", () => {
    const report = validateProjectModel(
      [
        {
          ...validSource,
          sourceType: "line",
          geometry: {
            type: "LineString",
            coordinates: [
              [0, 0],
              [1, 0],
            ],
          },
          properties: { source_acoustics_review_required: true },
        },
      ],
      [],
    );

    expect(
      report.warnings.some(
        (issue) => issue.code === "source.rls19.review_required",
      ),
    ).toBe(true);
  });

  it("stops warning once the reader has signed the acoustics off", () => {
    // The import's own flag stays: it records what OSM gave us, and a run
    // stamps it into provenance. What retires the finding is the second
    // property, the reader's sign-off — otherwise this is a warning no action
    // in the app can clear, and an OSM district import raises 608 of them.
    const report = validateProjectModel(
      [
        {
          ...validSource,
          sourceType: "line",
          geometry: {
            type: "LineString",
            coordinates: [
              [0, 0],
              [1, 0],
            ],
          },
          properties: {
            source_acoustics_review_required: true,
            source_acoustics_reviewed: true,
          },
        },
      ],
      [],
    );

    expect(
      report.warnings.some(
        (issue) => issue.code === "source.rls19.review_required",
      ),
    ).toBe(false);
  });

  it("rejects invalid RLS-19 source override values", () => {
    const report = validateProjectModel(
      [
        {
          ...validSource,
          sourceType: "line",
          geometry: {
            type: "LineString",
            coordinates: [
              [0, 0],
              [1, 0],
            ],
          },
          properties: {
            speed_pkw_kph: 0,
            traffic_day_pkw: -1,
            surface_type: "bogus",
          },
        },
      ],
      [],
    );

    expect(
      report.errors.some(
        (issue) => issue.code === "source.rls19.speed.invalid",
      ),
    ).toBe(true);
    expect(
      report.errors.some(
        (issue) => issue.code === "source.rls19.traffic.invalid",
      ),
    ).toBe(true);
    expect(
      report.errors.some(
        (issue) => issue.code === "source.rls19.surface_type.invalid",
      ),
    ).toBe(true);
  });
});

describe("RLS-19 Parkplatz validation", () => {
  const areaFeature = (properties: Record<string, unknown>): ModelFeature => ({
    id: "lot",
    kind: "source",
    sourceType: "area",
    properties,
    geometry: {
      type: "Polygon",
      coordinates: [
        [
          [0, 0],
          [10, 0],
          [10, 10],
          [0, 10],
          [0, 0],
        ],
      ],
    },
  });

  // This validator runs with no standard selected, and an area source is a
  // legitimate input to cnossos-industry and bub-industry too. Claiming every
  // one of them is a Parkplatz would bury those models in unrelated errors.
  it("leaves an area source carrying no parking property alone", () => {
    const report = validateProjectModel([areaFeature({})], []);

    expect(
      report.errors.filter((issue) =>
        issue.code.startsWith("source.rls19.parking."),
      ),
    ).toEqual([]);
  });

  it("refuses a half-filled Parkplatz, which is the realistic mistake", () => {
    const report = validateProjectModel(
      [areaFeature({ rls19_parking_num_spaces: 200 })],
      [],
    );

    const codes = report.errors.map((issue) => issue.code);
    expect(codes).toContain("source.rls19.parking.parking_type.missing");
    expect(codes).toContain("source.rls19.parking.movements.missing");
  });

  it("accepts a Tabelle 7 facility type in place of explicit rates", () => {
    const report = validateProjectModel(
      [
        areaFeature({
          rls19_parking_num_spaces: 200,
          rls19_parking_type: "lkw-omnibus",
          rls19_parking_facility_type: "tank-rastanlage",
        }),
      ],
      [],
    );

    expect(
      report.errors.filter((issue) =>
        issue.code.startsWith("source.rls19.parking."),
      ),
    ).toEqual([]);
  });

  it("keeps an explicitly stated zero movement rate legal", () => {
    const report = validateProjectModel(
      [
        areaFeature({
          rls19_parking_num_spaces: 200,
          rls19_parking_type: "pkw",
          rls19_parking_movements_per_space_day: 0.3,
          rls19_parking_movements_per_space_night: 0,
        }),
      ],
      [],
    );

    expect(
      report.errors.filter((issue) =>
        issue.code.startsWith("source.rls19.parking."),
      ),
    ).toEqual([]);
  });

  const square = (offset: number): [number, number][][] => [
    [
      [offset, 0],
      [offset + 10, 0],
      [offset + 10, 10],
      [offset, 10],
      [offset, 0],
    ],
  ];

  const complete = {
    rls19_parking_num_spaces: 200,
    rls19_parking_type: "pkw",
    rls19_parking_movements_per_space_day: 0.3,
    rls19_parking_movements_per_space_night: 0,
  };

  // Every property is filled in, so nothing else here has anything to say — and
  // until this rule existed that was the whole finding: a lot a user could
  // complete on the map, save, and have refused only when a run read it.
  it("refuses a multi-part Parkplatz the extractors cannot read", () => {
    const feature = areaFeature(complete);
    feature.geometry = {
      type: "MultiPolygon",
      coordinates: [square(0), square(50)],
    };

    const report = validateProjectModel([feature], []);

    const issue = report.errors.find(
      (candidate) =>
        candidate.code === "source.rls19.parking.geometry.multipart",
    );
    // The part count is the finding's payload and is asserted here; that it
    // reaches a sentence naming the Teilfläche is `validation-message.test.ts`'s
    // business, because the sentence is now per locale and this file is not.
    expect(issue?.params).toEqual({
      parts: 2,
      field: "rls19_parking_num_spaces",
    });
  });

  it("accepts a MultiPolygon carrying exactly one part, as the extractors do", () => {
    const feature = areaFeature(complete);
    feature.geometry = { type: "MultiPolygon", coordinates: [square(0)] };

    const report = validateProjectModel([feature], []);

    expect(
      report.errors.filter((issue) =>
        issue.code.startsWith("source.rls19.parking."),
      ),
    ).toEqual([]);
  });
});

/**
 * The gap that let an OSM import be shown as clean and then refused.
 *
 * `validate.ts` had no geometry checks at all, so every
 * `geometry.*.self_intersection` the backend produces was invisible here — 45
 * of them across the 2397 ways of a Berlin extract, every one of which the
 * preview called clean and the save then had an opinion about.
 *
 * The severity split is the backend's, not a second opinion — a ring's winding
 * is read by point-in-polygon and the screening crossing counts, a line's is
 * read by nothing.
 */
describe("self-intersection", () => {
  const bowTie: [number, number][] = [
    [0, 0],
    [10, 10],
    [10, 0],
    [0, 10],
    [0, 0],
  ];

  it("refuses a footprint whose ring crosses itself", () => {
    const report = validateProjectModel(
      [
        {
          id: "bld-x",
          kind: "building",
          heightM: 10,
          geometry: { type: "Polygon", coordinates: [bowTie] },
        },
      ],
      [],
    );

    expect(report.valid).toBe(false);
    expect(report.errors.map((issue) => issue.code)).toContain(
      "geometry.polygon.self_intersection",
    );
  });

  it("keeps a line that crosses itself, and says so", () => {
    const report = validateProjectModel(
      [
        {
          id: "src-x",
          kind: "source",
          sourceType: "line",
          geometry: {
            type: "LineString",
            coordinates: [
              [0, 0],
              [10, 10],
              [0, 10],
              [10, 0],
            ],
          },
        },
      ],
      [],
    );

    // A road drawn as one way that touches itself is a roundabout. Refusing it
    // made an OSM extract unsaveable over geometry it holds correctly.
    expect(report.valid).toBe(true);
    expect(report.warnings.map((issue) => issue.code)).toContain(
      "geometry.linestring.self_intersection",
    );
  });

  it("answers the same for one footprint in degrees and in metres", () => {
    // The check used to compare a cross product — an area, so the coordinate
    // unit squared — against a fixed tolerance, so the same building was
    // self-intersecting in EPSG:4326 and simple in EPSG:25832.
    const shape: [number, number][] = [
      [0, 0],
      [3, 0],
      [3, 1],
      [1, 1],
      [1, 3],
      [0, 3],
      [0, 0],
    ];
    const rotation = (31 * Math.PI) / 180;

    const ring = (
      unit: number,
      originX: number,
      originY: number,
    ): [number, number][] =>
      shape.map(([px, py]) => [
        originX + (px * Math.cos(rotation) - py * Math.sin(rotation)) * unit,
        originY + (px * Math.sin(rotation) + py * Math.cos(rotation)) * unit,
      ]);

    const inDegrees = ring(1e-5, 13.38, 52.51);
    const inMetres = ring(1e-5 * 111320, 390000, 5819000);

    for (const [name, coordinates] of [
      ["degrees", inDegrees],
      ["metres", inMetres],
    ] as const) {
      const report = validateProjectModel(
        [
          {
            id: `bld-${name}`,
            kind: "building",
            heightM: 10,
            geometry: { type: "Polygon", coordinates: [coordinates] },
          },
        ],
        [],
      );

      expect(report.errors, name).toEqual([]);
    }
  });

  it("reports the cost bound rather than walking a huge geometry", () => {
    // Above the limit the quadratic walk is not attempted. In a browser a
    // frozen tab is worse than an unchecked geometry, so it is reported.
    const coordinates: [number, number][] = [];
    for (let i = 0; i <= SELF_INTERSECTION_POINT_LIMIT; i++) {
      coordinates.push([i % 2, i]);
    }

    const report = validateProjectModel(
      [
        {
          id: "src-long",
          kind: "source",
          sourceType: "line",
          geometry: { type: "LineString", coordinates },
        },
      ],
      [],
    );

    expect(report.valid).toBe(true);
    expect(report.warnings.map((issue) => issue.code)).toContain(
      "geometry.linestring.self_intersection.skipped",
    );
  });
});

File: backend/internal/geo/modelgeojson/validate_test.go
package modelgeojson

import (
	"math"
	"strings"
	"testing"
)

func TestNormalizeAndValidateValidModel(t *testing.T) {
	t.Parallel()

	payload := []byte(`{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {"id": "src-1", "kind": "source", "source_type": "point"},
      "geometry": {"type": "Point", "coordinates": [350000.0, 5800000.0]}
    },
    {
      "type": "Feature",
      "properties": {"id": "b-1", "kind": "building", "height_m": 12.0},
      "geometry": {"type": "Polygon", "coordinates": [[[0,0],[10,0],[10,10],[0,10],[0,0]]]}
    },
    {
      "type": "Feature",
      "properties": {"id": "bar-1", "kind": "barrier", "height_m": 2.5},
      "geometry": {"type": "LineString", "coordinates": [[1,1],[5,1],[8,2]]}
    },
    {
      "type": "Feature",
      "properties": {"id": "rcv-1", "kind": "receiver", "height_m": 4.0},
      "geometry": {"type": "Point", "coordinates": [20.0, 30.0]}
    }
  ]
}`)

	model, err := Normalize(payload, "EPSG:25832", "input.geojson")
	if err != nil {
		t.Fatalf("normalize: %v", err)
	}

	report := Validate(model)
	if !report.Valid {
		t.Fatalf("expected valid model, got errors: %#v", report.Errors)
	}

	if report.ErrorCount() != 0 {
		t.Fatalf("expected 0 errors, got %d", report.ErrorCount())
	}

	if report.WarningCount() != 0 {
		t.Fatalf("expected 0 warnings, got %d", report.WarningCount())
	}

	if model.ToDump().FeatureCount != 4 {
		t.Fatalf("expected 4 features in dump")
	}
}

func TestValidateReceiverRequiresPositiveHeightAndPointGeometry(t *testing.T) {
	t.Parallel()

	payload := []byte(`{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {"id": "rcv-1", "kind": "receiver", "height_m": 0},
      "geometry": {"type": "LineString", "coordinates": [[0,0],[1,1]]}
    }
  ]
}`)

	model, err := Normalize(payload, "EPSG:25832", "input.geojson")
	if err != nil {
		t.Fatalf("normalize: %v", err)
	}

	report := Validate(model)
	if report.Valid {
		t.Fatal("expected invalid receiver feature")
	}

	foundHeight := false
	foundGeometry := false

	for _, issue := range report.Errors {
		if issue.Code == "receiver.height.invalid" {
			foundHeight = true
		}

		if issue.Code == "receiver.geometry.invalid" {
			foundGeometry = true
		}
	}

	if !foundHeight || !foundGeometry {
		t.Fatalf("expected receiver validation errors, got %#v", report.Errors)
	}
}

func TestValidateMissingBuildingHeight(t *testing.T) {
	t.Parallel()

	payload := []byte(`{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {"id": "b-1", "kind": "building"},
      "geometry": {"type": "Polygon", "coordinates": [[[0,0],[10,0],[10,10],[0,10],[0,0]]]}
    }
  ]
}`)

	model, err := Normalize(payload, "EPSG:4326", "input.geojson")
	if err != nil {
		t.Fatalf("normalize: %v", err)
	}

	report := Validate(model)
	if report.Valid {
		t.Fatal("expected validation error")
	}

	if report.ErrorCount() == 0 {
		t.Fatal("expected at least one error")
	}
}

func TestValidateSelfIntersectingPolygon(t *testing.T) {
	t.Parallel()

	payload := []byte(`{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {"id": "b-2", "kind": "building", "height_m": 10},
      "geometry": {"type": "Polygon", "coordinates": [[[0,0],[10,10],[10,0],[0,10],[0,0]]]}
    }
  ]
}`)

	model, err := Normalize(payload, "EPSG:4326", "input.geojson")
	if err != nil {
		t.Fatalf("normalize: %v", err)
	}

	report := Validate(model)
	if report.Valid {
		t.Fatal("expected invalid model")
	}

	found := false

	for _, issue := range report.Errors {
		if issue.Code == "geometry.polygon.self_intersection" {
			found = true
			break
		}
	}

	if !found {
		t.Fatalf("expected self-intersection error, got %#v", report.Errors)
	}
}

func TestValidateProjectedCRSWithLonLatWarning(t *testing.T) {
	t.Parallel()

	payload := []byte(`{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {"id": "src-2", "kind": "source", "source_type": "point"},
      "geometry": {"type": "Point", "coordinates": [13.4, 52.5]}
    }
  ]
}`)

	model, err := Normalize(payload, "EPSG:25832", "input.geojson")
	if err != nil {
		t.Fatalf("normalize: %v", err)
	}

	report := Validate(model)
	if report.ErrorCount() != 0 {
		t.Fatalf("expected no errors, got %#v", report.Errors)
	}

	if report.WarningCount() == 0 {
		t.Fatal("expected at least one warning")
	}
}

func TestNormalizePreservesCustomPropertiesRoundTrip(t *testing.T) {
	t.Parallel()

	payload := []byte(`{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {
        "id": "src-3",
        "kind": "source",
        "source_type": "line",
        "road_surface_type": "concrete",
        "traffic_day_light_vph": 123.5
      },
      "geometry": {"type": "LineString", "coordinates": [[0,0],[10,0]]}
    }
  ]
}`)

	model, err := Normalize(payload, "EPSG:25832", "input.geojson")
	if err != nil {
		t.Fatalf("normalize: %v", err)
	}

	if got := model.Features[0].Properties["road_surface_type"]; got != "concrete" {
		t.Fatalf("expected preserved road_surface_type, got %#v", got)
	}

	collection := model.ToFeatureCollection()

	gotProps := collection.Features[0].Properties
	if got := gotProps["road_surface_type"]; got != "concrete" {
		t.Fatalf("expected round-tripped road_surface_type, got %#v", got)
	}

	if got := gotProps["traffic_day_light_vph"]; got != 123.5 {
		t.Fatalf("expected round-tripped traffic_day_light_vph, got %#v", got)
	}
}

// hasErrorCode reports whether the report carries an error with that code, and
// the feature ids it was raised against.
func hasErrorCode(report ValidationReport, code string) (string, bool) {
	for _, issue := range report.Errors {
		if issue.Code == code {
			return issue.FeatureID, true
		}
	}

	return "", false
}

func validateCalcAreaPayload(t *testing.T, payload string) ValidationReport {
	t.Helper()

	model, err := Normalize([]byte(payload), "EPSG:25832", "input.geojson")
	if err != nil {
		t.Fatalf("normalize: %v", err)
	}

	return Validate(model)
}

func TestValidateCalcAreaPolygonWithoutHeightIsValid(t *testing.T) {
	t.Parallel()

	report := validateCalcAreaPayload(t, `{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {"id": "calc-1", "kind": "calc-area"},
      "geometry": {"type": "Polygon", "coordinates": [[[350000,5800000],[350100,5800000],[350100,5800100],[350000,5800100],[350000,5800000]]]}
    }
  ]
}`)

	if !report.Valid {
		t.Fatalf("expected a calc-area polygon without height_m to validate, got %#v", report.Errors)
	}
}

func TestValidateCalcAreaRejectsNonPolygonGeometry(t *testing.T) {
	t.Parallel()

	geometries := map[string]string{
		"LineString":   `{"type": "LineString", "coordinates": [[350000,5800000],[350100,5800100]]}`,
		"Point":        `{"type": "Point", "coordinates": [350000,5800000]}`,
		"MultiPolygon": `{"type": "MultiPolygon", "coordinates": [[[[350000,5800000],[350100,5800000],[350100,5800100],[350000,5800100],[350000,5800000]]]]}`,
	}

	for name, geometry := range geometries {
		t.Run(name, func(t *testing.T) {
			t.Parallel()

			report := validateCalcAreaPayload(t, `{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {"id": "calc-1", "kind": "calc-area"},
      "geometry": `+geometry+`
    }
  ]
}`)

			featureID, found := hasErrorCode(report, "calcarea.geometry.invalid")
			if !found {
				t.Fatalf("expected calcarea.geometry.invalid for %s, got %#v", name, report.Errors)
			}

			if featureID != "calc-1" {
				t.Fatalf("expected the error to name calc-1, got %q", featureID)
			}
		})
	}
}

func TestValidateRejectsASecondCalcArea(t *testing.T) {
	t.Parallel()

	report := validateCalcAreaPayload(t, `{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {"id": "calc-1", "kind": "calc-area"},
      "geometry": {"type": "Polygon", "coordinates": [[[350000,5800000],[350100,5800000],[350100,5800100],[350000,5800100],[350000,5800000]]]}
    },
    {
      "type": "Feature",
      "properties": {"id": "calc-2", "kind": "calc-area"},
      "geometry": {"type": "Polygon", "coordinates": [[[360000,5800000],[360100,5800000],[360100,5800100],[360000,5800100],[360000,5800000]]]}
    }
  ]
}`)

	featureID, found := hasErrorCode(report, "model.calc_area.duplicate")
	if !found {
		t.Fatalf("expected model.calc_area.duplicate, got %#v", report.Errors)
	}

	// The second one is the one that can be removed, so it is the one named.
	if featureID != "calc-2" {
		t.Fatalf("expected the duplicate error to name calc-2, got %q", featureID)
	}
}

func TestValidateUnknownKindMessageListsEveryKind(t *testing.T) {
	t.Parallel()

	report := validateCalcAreaPayload(t, `{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {"id": "x-1", "kind": "calculation-area"},
      "geometry": {"type": "Point", "coordinates": [350000,5800000]}
    }
  ]
}`)

	var message string

	for _, issue := range report.Errors {
		if issue.Code == "feature.kind.invalid" {
			message = issue.Message
		}
	}

	if message == "" {
		t.Fatalf("expected feature.kind.invalid, got %#v", report.Errors)
	}

	for _, kind := range FeatureKinds {
		if !strings.Contains(message, kind) {
			t.Fatalf("kind rejection message %q does not mention %q", message, kind)
		}
	}
}

// A road drawn as one way that touches itself is a roundabout, not a defect,
// and nothing in the model reads a line's winding — so it is reported and the
// model still saves. 45 of 2397 ways in a Berlin OSM extract are like this.
func TestValidateSelfIntersectingLineIsAWarning(t *testing.T) {
	t.Parallel()

	payload := []byte(`{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {"id": "s-1", "kind": "source", "source_type": "line"},
      "geometry": {"type": "LineString", "coordinates": [[0,0],[10,10],[0,10],[10,0]]}
    }
  ]
}`)

	model, err := Normalize(payload, "EPSG:4326", "input.geojson")
	if err != nil {
		t.Fatalf("normalize: %v", err)
	}

	report := Validate(model)
	if !report.Valid {
		t.Fatalf("a self-intersecting line must not refuse the model, got %#v", report.Errors)
	}

	if !hasCode(report.Warnings, "geometry.linestring.self_intersection") {
		t.Fatalf("expected a self-intersection warning, got %#v", report.Warnings)
	}
}

// The same footprint must get the same answer whichever CRS it arrives in.
//
// It did not. orientation compared a cross product — an area, so the coordinate
// unit squared — against a fixed 1e-9. That is large next to two short edges in
// degrees (~1e-10) and negligible next to the same two in metres (~1e0), so the
// predicate answered "collinear" in one CRS and "turning" in the other. A false
// "collinear" then falls through to onSegment, which is a bounding-box test and
// sound only when the triple really is collinear, and any two edges of a
// non-convex footprint whose boxes overlap read as crossing.
//
// The shape below is a rotated L: small, because a synthetic reproducer needs
// short edges to get the cross product down there. Real OSM footprints reach
// the same place at ordinary sizes by having many more vertices — it flagged
// 219 of 631 buildings in a Berlin extract, every one of them a simple polygon,
// and that is why an OSM import could not be saved at all.
func TestValidateSelfIntersectionIsScaleInvariant(t *testing.T) {
	t.Parallel()

	shape := [][2]float64{{0, 0}, {3, 0}, {3, 1}, {1, 1}, {1, 3}, {0, 3}, {0, 0}}

	const rotation = 31 * math.Pi / 180

	ring := func(unit, originX, originY float64) []point2 {
		out := make([]point2, 0, len(shape))
		for _, p := range shape {
			x := p[0]*math.Cos(rotation) - p[1]*math.Sin(rotation)
			y := p[0]*math.Sin(rotation) + p[1]*math.Cos(rotation)
			out = append(out, point2{x: originX + x*unit, y: originY + y*unit})
		}

		return out
	}

	const degreeUnit = 1e-5

	for _, tc := range []struct {
		name string
		ring []point2
	}{
		// EPSG:4326 near Berlin, and the same footprint in EPSG:25832 metres.
		{name: "degrees", ring: ring(degreeUnit, 13.38, 52.51)},
		{name: "metres", ring: ring(degreeUnit*111320, 390000, 5819000)},
	} {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()

			if hasSelfIntersection(tc.ring, true) {
				t.Errorf("a simple L-shaped footprint was reported as self-intersecting in %s", tc.name)
			}
		})
	}
}

// A genuine bow-tie must still be caught at building scale, where the whole
// ring spans a ten-thousandth of a degree — the relative tolerance has to be
// tight enough to see a crossing that small.
func TestValidateSelfIntersectionStillCaughtAtBuildingScale(t *testing.T) {
	t.Parallel()

	const s = 1e-4

	ring := []point2{
		{x: 13.38, y: 52.51},
		{x: 13.38 + s, y: 52.51 + s},
		{x: 13.38 + s, y: 52.51},
		{x: 13.38, y: 52.51 + s},
		{x: 13.38, y: 52.51},
	}

	if !hasSelfIntersection(ring, true) {
		t.Error("a bow-tie footprint the size of a real building was not detected")
	}
}

func hasCode(issues []ValidationIssue, code string) bool {
	for _, issue := range issues {
		if issue.Code == code {
			return true
		}
	}

	return false
}
